src/Components/Components/src/Routing/Router.cs

```diff
@@ -75,7 +76,20 @@ public class Router : IComponent, IHandleAfterRender, IDisposable
         /// </summary>
         [Parameter] public EventCallback<NavigationContext> OnNavigateAsync { get; set; }
 
-        private RouteTable Routes { get; set; }
+        /// <summary>
+        /// Gets or sets a flag to indicate whether route matching should prefer exact matches
+        /// over wildcards.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// Important: all applications should explicitly set this to true. The option to set it to false
+        /// (or leave unset, which defaults to false) is only provided for backward compatibility.
+        /// In .NET 6, this option will be removed and the router will always prefer exact matches.
+        /// </para>
+        /// </remarks>
+        [Parameter] public bool PreferExactMatches { get; set; }
```

Comment:
After consideration, I think we should not name the new option after the package version number, because:

1. This is going in a patch, so the version would be something like `RoutingBehavior_5_0_1` which is ridiculously ugly to put in the default template
2. We can’t really use a name like `Current` because theoretically there could be an even newer incompatible version in the future

My proposal here of calling it `PreferExactMatches` has the benefits:

* It doesn't look particularly unpleasant in the default template. It sounds like a good thing, which it is.
* Even after we retire the option in 6.0 and remove it from intellisense, we can continue to allow people to pass  `PreferExactMatches="true"` (that just becomes a no-op, whereas `PreferExactMatches="false"` becomes an error). It doesn't make anyone's code look wrong to have this still in it, plus it doesn't imply the person *only* wants 5.0.1 behaviors and not newer ones
* If we ever need to fork the logic again, we can use the same pattern of naming the new option after its own behavior, and that's independent of this.

Reply:
I would actually prefer that we only have a single flag for this.

I would hope we can avoid having to deal with a cross-section of things we can turn on/off and make it a multi value instead or something different that only allows you to set compatibility to a point in time.

Reply:
As long as we don't end up with more incompatible routing algorithm changes in 5.0.x, then there will only be a single flag. It's going away immediately in 6.0 - we're not carrying any complexity forwards into the long term with it.

Reply:
Fair enough